test/test_train_imagenet.py

```diff
@@ -180,12 +203,16 @@ def test_loop_fn(model, loader, device, context):
 
   accuracy = 0.0
   writer = SummaryWriter(log_dir=FLAGS.logdir) if FLAGS.logdir else None
+  num_training_steps_per_epoch = len(train_dataset.imgs) // (
+      FLAGS.batch_size * (len(devices) or 1))
```

Comment:
is this true for distributed training? i.e. is `train_dataset` sharded at this point?

Reply:
@jysohn23 is there a way to compute this number in the distributed case?

Reply:
No wait. Let me expose the replication devices explicitly instead of doing that.

Reply:
Without API ... I'd use `xm.xla_replication_devices(devices)` where `devices` are the devices you pass to `DataParallel`.

Reply:
IIRC something like that was already there, but we ended up getting rid of it, I can't recall why.

A simpler solution would be to have train_loop fn return `x+1`  (I'd rename that too)

```
for x, (data, target) in loader:
```

at the end of the loop, then add all the x's from various devices, that's the total number of steps this machine did in the `model_parallel` call w/o computing or relying on anything.

Reply:
What?

Reply:
I mean this:

```python

def train_loop_fn(...):
  ...
  for step, data in loader:
    ...
  return step + 1

steps = model_parallel(train_loop_fn, loader)
num_training_steps_per_epoch = sum(steps)
global_step = (epoch -1) * num_training_steps_per_epoch
```

Reply:
But how can I compute the learning rate during the warmup epoch without knowing the denominator?

Reply:
Ah I see this is used inside `train_loop_fn`. Reading this part of code, I thought its use was updating global_step only.

Reply:
Chicken/Egg problem :smile: 
Use `len(xm.xla_replication_devices(devices))` ... if `len(devices) > 1` else `1`.
Note that replication is valid only in case of one device, or all devices.

Reply:
Switched to Davide's suggestion